numpy/_core/tests/test_multiarray.py

```diff
@@ -8492,6 +8492,23 @@ def __array__(self, dtype=None):
                                               "and 'copy' keywords")):
             np.array(a, copy=False)
 
+    @pytest.mark.skipif(not HAS_REFCOUNT, reason="Python lacks refcounts")
+    def test__array__reference_leak(self):
+        class NotAnArray:
+            def __array__(self):
+                raise NotImplementedError()
+
+        x = NotAnArray()
+
+        refcount = sys.getrefcount(x)
+
+        try:
+            np.array(x)
+        except NotImplementedError:
+            pass
+
```

Comment:
Do we need `gc.collect()` here to prevent test flakiness?

Reply:
It doesn't look like we do that elsewhere in this file but can't hurt.

Reply:
There are no reference cycles possible here, so I don't think it matters either way.